packages/components/src/input-control/state.js

```diff
@@ -0,0 +1,253 @@
+/**
+ * External dependencies
+ */
+import { flattenDeep } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import { useReducer } from '@wordpress/element';
+
+const initialStateReducer = ( state ) => state;
+
+const initialInputControlState = {
+	_event: {},
+	error: null,
+	initialValue: '',
+	isDirty: false,
+	isDragEnabled: false,
+	isDragging: false,
+	isPressEnterToChange: false,
+	value: '',
+};
+
+const actionTypes = {
+	CHANGE: 'CHANGE',
+	COMMIT: 'COMMIT',
+	DRAG_END: 'DRAG_END',
+	DRAG_START: 'DRAG_START',
+	DRAG: 'DRAG',
+	INVALIDATE: 'INVALIDATE',
+	PRESS_DOWN: 'PRESS_DOWN',
+	PRESS_ENTER: 'PRESS_ENTER',
+	PRESS_UP: 'PRESS_UP',
+	RESET: 'RESET',
+	UPDATE: 'UPDATE',
+};
+
+export const inputControlActionTypes = actionTypes;
+
+/**
+ * Prepares initialState for the reducer.
+ *
+ * @param {Object} initialState The initial state.
+ * @return {Object} Prepared initialState for the reducer
+ */
+function mergeInitialState( initialState = initialInputControlState ) {
+	const { value } = initialState;
+
+	return {
+		...initialInputControlState,
+		...initialState,
+		initialValue: value,
+	};
+}
+
+/**
+ * Combines multiple stateReducers into a single stateReducer, building
+ * the pipeline to control the flow for state and actions.
+ *
+ * @param  {...any} fns State reducers.
+ * @return {Function} The single combined stateReducer.
+ */
+export const combineStateReducers = ( ...fns ) => {
+	// Using lodash.flattenDeep until we have an Array.flat polyfill.
+	const reducers = flattenDeep( fns );
```

Comment:
Should we keep this flattening? Do you have concrete examples of when you'd want to traverse a deep list of reducers? Otherwise we might keep a stricter interface and only take `...fns` with `fns :: Array<Reducer>`.

I think being strict about this idea that reducers are _linear input_ rather than tree-like is also more consistent with my above comment on `composeReducers`. Saves us and API consumers from confusion. :)

Reply:
Ah, good catch! Yes, since we're explicitly using it between the layers (example below), we no longer need to flatten the reducer args.

```jsx
<Input
	__unstableStateReducer={ combineStateReducers(
		numberControlStateReducer,
		stateReducer
	) }
/>
```